src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/prisonperson/PrisonPersonMigrationService.kt

```diff
@@ -25,6 +34,41 @@ class PrisonPersonMigrationService(
     )
   }
 
+  @Transactional
+  suspend fun upsert(mappingRequest: PrisonPersonMigrationMappingRequest): PrisonPersonMigrationMapping =
+    mappingRequest.find()
+      ?.let { mappingRequest.update() }
+      ?: let { mappingRequest.insert() }
+
+  private suspend fun PrisonPersonMigrationMappingRequest.find() =
+    repository.findByNomisPrisonerNumberAndMigrationTypeAndLabel(nomisPrisonerNumber, migrationType, label)
+
+  private suspend fun PrisonPersonMigrationMappingRequest.update(): PrisonPersonMigrationMapping =
+    template.update<PrisonPersonMigrationMapping>()
+      .inTable("prison_person_migration_mapping")
+      .matching(
+        query(
+          where("nomis_prisoner_number").`is`(nomisPrisonerNumber)
+            .and(where("migration_type").`is`(migrationType))
+            .and(where("label").`is`(label)),
+        ),
+      )
+      .apply(update("dps_ids", dpsIds.toString()))
+      .awaitSingle()
+      .let { find()!! }
+
+  private suspend fun PrisonPersonMigrationMappingRequest.insert(): PrisonPersonMigrationMapping =
+    template.insert<PrisonPersonMigrationMapping>()
+      .using(
+        PrisonPersonMigrationMapping(
+          nomisPrisonerNumber = nomisPrisonerNumber,
+          migrationType = migrationType,
+          dpsIds = dpsIds.toString(),
+          label = label,
+        ),
+      )
+      .awaitSingle()
```

Comment:
Spring data R2DBC still [doesn't support composite entity keys](https://github.com/spring-projects/spring-data-relational/issues/574) so we needed another way. I thought I'd try out the [template](https://docs.spring.io/spring-data/relational/reference/r2dbc/entity-persistence.html#r2dbc.entityoperations.fluent-api.insert) instead.